### ISSUE TITLE
Change apiserver healthiness check in KCM

### ISSUE THREAD

**Issue by jyotimahapatra** [kubernetes/kubernetes#108014]
### What happened?

KCM startup has a dependency on apiserver availability https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/controller-manager/app/helper.go#L37.

When we have an apiserver that uses kms keys, an error in the kms plugin, the kms service, or a customer mistake of deleting the key interrupt the  kube-controller-manager startup because of the dependency on healthz.

### What did you expect to happen?

Would it be ok to introduce a flag on the kube-controller-manager options to exclude certain healthz checks? In such a case, kcm would be able to start up even when kms has an issue. Not all the controller loops will be able to startup in that situation, but it might be better to have the kcm be able to start and try to reconcile the loops.

### How can we reproduce it (as minimally and precisely as possible)?

Create an EKS cluster using kms enabled. 
```
aws eks 
    --role-arn arn:aws:iam::xxx:role/xxx \
    --resources-vpc-config subnetIds=subnet-xxx,subnet-xxx,securityGroupIds=sg-xxx \
    --kubernetes-version 1.19 --encryption-config resources=secrets,provider={keyArn=arn:aws:kms:us-west-2:xxx:key/xxxx}
```

Disable the kms key. The kube controller manager becomes unhealthy

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>
EKS
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
this iissue is a good example on why I think the controller manager should not start if the apiserver is not healthy https://github.com/kubernetes/kubernetes/issues/107583

**Comment by jyotimahapatra** [comment]
checks on apiserver unhealthy is definitely good. And hence the question about excluding certain checks like kms through a flag. When kcm is up , apiserver unhealthiness does not make it shut down. Following that, it might be ok to let the kcm start under certain hc failures and let the controllers recover when apiserver comes back. The kcm healthiness could also be more specific about which controllers are failing.

**Comment by logicalhan** [comment]
> this issue is a good example on why I think the controller manager should not start if the apiserver is not healthy #107583

Yeah I agree with you, I'm not generally in favor of removing the apiserver healthiness check for reasons exactly like this. 

> Not all the controller loops will be able to startup in that situation, but it might be better to have the kcm be able to start and try to reconcile the loops.

Why is it a problem that the controller manager is restarting in this situation? What problem are you trying to fix?

**Comment by jyotimahapatra** [comment]
> Yeah I agree with you, I'm not generally in favor of removing the apiserver healthiness check for reasons exactly like this.

I agree and I'm not advocating a removal. Im looking for a way to exclude certain kind of health check.

> Why is it a problem that the controller manager is restarting in this situation? What problem are you trying to fix?

The issue I'm facing is certainly not a widespread issue. We are operating a platform for customers to create clusters on and we apply the general configs. Certain clusters do use kms encryption. Cluster admins can delete/disable kms keys and make the cluster unhealthy. From a platform team perspective, we look at the health of all clusters and page when they are unhealthy. 
When such cluster admin related errors happen, we dont want to page the platform team since there is no action needed. Right now we are able to prevent the noise in apiserver by excluding kms check in apiserver. Although that keeps the apiserver up and running, users wont be able to create new secrets and the error message will help them debug. 
However, the KCM still pages us when its unhealthy and in many scenarios these pages are non actionable because an admin has deleted the kms key. We are looking for a way to reduce the noise by adding a flag (only on clusters using kms) that will exclude kms `/healthz?exclude=kms-provider-0` .

This change will help smoothen the operations and reduce pages for platform teams.

**Comment by liggitt** [comment]
we discussed this in sig-auth on 4/27 where this topic showed up first (and then routed it to sig-api-machinery)

a few thoughts from there:
* why does kube-controller-manager need to check all all of healthz/readyz? I remember kcm requiring discovery info, but not necessarily all of healthz/readyz. can/should we scope down what kcm waits for to just discovery info?
* what does kcm do if the healthz check passes, kcm continues startup, and the health check later starts failing? does it respond correctly?

I'm skeptical about adding more opt-in/opt-out readiness knobs to kcm. I'd rather see us scope down what kcm considers to the minimum required than pay attention to too much and let individual deployers opt out of things they think are optional.

**Comment by lavalamp** [comment]
* most KCM installs just have it talking to the local apiserver. If that apiserver is unhealthy but others are healthy, it makes sense to give up the lock (you might not even have a choice depending on how unhealthy apiserver is).
* If for some reason KCM is going through a load balancer to get to apiserver, then it doesn't really make sense to try and lose the lock, since you're probably not talking to a specific apiserver and no other KCM would do better anyway.
* In this particular case, since KMS health requires user action, what I think you really want is to globally exclude it from apiserver health checks, not just for KCM.
* In the abstract, I'm not in favor of adding knobs to KCM, due to my first two points implying that different behavior is needed in different setups, which implies yet more flags... Let's not start down this road until it's certain we need it.

I'm not sure how I feel about the use case, I think a better solution might be to somehow lock the key so the user can't delete it while it's in use, only rotate it.

**Comment by jyotimahapatra** [comment]
> why does kube-controller-manager need to check all all of healthz/readyz? I remember kcm requiring discovery info, but not necessarily all of healthz/readyz. can/should we scope down what kcm waits for to just discovery info?
Certainly, I don't think there is merit in stopping the KCM startup due to apiserver unhealthiness, more so when there is no such check at runtime. The behavior should be rather consistent.

> what does kcm do if the healthz check passes, kcm continues startup, and the health check later starts failing? does it respond correctly?
It depends on what is really failing. One of the apiserver health check type is `log` and that should not impact KCM. etcd unhealthiness would certainly impact. Each controller could fail diifferently due to apiserver unhealthiness. I think there is no standard way for controllers to communicate such failures.

> what I think you really want is to globally exclude it from apiserver health checks, not just for KCM.
In my infra, this is already taken care of from all places that depend on apiserver health check. The system becomes statically stable when `log` or `kms` checks fail. However, which the current hardcoding in KCM, i'm not able to stablilize the system when a non impacting event happens.

I agree adding more flags is not the best course. Removing the health check seems better to me.

**Comment by jyotimahapatra** [comment]
Doing some code archaeology, the change was done in #60291 for fixing #60288 . The change was done with intent `We should deprecate or remove ServerAPIVersions and make use of /healthz as all other components do.` . This consistency is good, but introduced strict check on apiserver to be available before KCM can even start. I lack context, but i think it is not necessary to have this guarantee. The controller loops should be resilient to apiserver being unhealthy, more so since apiserver can become unhealthy later after KCM starts.

An alternative approach: If discovery is the only requirement, would it be ok to include healthz for just the etcd? 

I need to experiment this: say, we remove the requirement of healthz altogether, what does the KCM bootstrap behavior look like?

**Comment by liggitt** [comment]
> The controller loops should be resilient to apiserver being unhealthy, more so since apiserver can become unhealthy later after KCM starts.

I agree that is a goal, but until we are sure that the controller loops *are* resilient to apiserver being unhealthy, removing the `/healthz` check exposes controllers to API servers that have not completed one-time operations that happen at API server start (which inform `/healthz`).

All the `poststarthook` subchecks indicate one-time startup operations have completed, which will not go unhealthy after they complete successfully.

The `autoregister-completion` subcheck is also a one-time startup operation that indicates the APIServices for the expected served API groups have completed being created successfully.

To do this safely, we'd have to look at how all the controllers use discovery information, and make sure none of the controllers make startup-time decisions based on discovery information that would be dangerous or leave controller loops in a degraded or non-functional state.

**Comment by jyotimahapatra** [comment]
I ran a patch in production with kms provider turned off https://github.com/kubernetes/kubernetes/pull/116976/files . Any controller that depended on discovery (eg. garbage collection controller, resource quota controller) were not able to operate. All other controllers don't depend on KMS and were able to operate just fine. As a benefit, when users deleted KMS keys accidentally, they were still able to salvage the cluster. Without the change, the kube-controller-manager summarily stops randomly (depending on when kcm restarts) after KMS key is deleted.
This change has been deployed to EKS production and has helped users salvage production clusters when they delete KMS keys by mistake. 

My understanding is that with the proposed changes we will let controller manager skip only the KMS related health checks and not everything. If a post start hook fails, the controller manager behavior stays unchanged. If everything else is set correctly and kms key is deleted, the change helps users salvage a cluster.

I have used this change in thousands of clusters and can state that controller manager loops are resilient to be able to reconcile back when key is fixed.
